gnrc_ipv6_nib: provide internal implementation for off-link entries

miri64 opened the pull request:
Implementation off the off-link component of the NIB back-end.

I needed to move some stuff around, to keep consistency, but other than that the change should be straight forward.

This PR is part of the network layer remodelling effort:
![PR dependencies](https://miri64.github.io/riot-ndp-model/PR%20overview.svg)

miri64 commented on sys/include/net/gnrc/ipv6/nib/conf.h:
```diff
@@ -184,6 +184,16 @@ extern "C" {
+/**
+ * @brief   Number of off-link entries in NIB
+ *
+ * @note    **This number has direct influence on the maximum number of
```
Tried to keep the style of the original ones.

smlng replied:
I understand - and docu styling might be separate more general issue anyway 😉

miri64 replied:
Will fix in a different PR ;-)

miri64 replied:
@smlng see https://github.com/RIOT-OS/RIOT/pull/7218

cgundogan commented on sys/include/net/gnrc/ipv6/nib/conf.h:
```diff
@@ -184,6 +184,16 @@ extern "C" {
+/**
+ * @brief   Number of off-link entries in NIB
+ *
+ * @attention   This number has direct influence on the maximum number of
```
just to clarify: how exactly does `GNRC_IPV6_NIB_OFFL_NUMOF` influences the max. number? I would assume it **is** the max. number, or is there another number that is added to that?

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c:
```diff
@@ -62,14 +64,13 @@ _nib_onl_entry_t *_nib_onl_alloc(const ipv6_addr_t *addr, unsigned iface)
 
         if ((_nib_onl_get_if(tmp) == iface) &&
-            (ipv6_addr_equal(addr, &tmp->ipv6))) {
+            ((addr == NULL) || (ipv6_addr_equal(addr, &tmp->ipv6)))) {
```
why do you need this check for `addr == NULL` here? There's a `@pre` in the doxygen of this function, that `addr != NULL`.

cgundogan replied:
you probably need to remove the doxygen `@pre` then.

miri64 replied:
The `@pre` is wrong. For prefix list entries the next-hop address may be `NULL`.

miri64 replied:
Found an error here, while debugging NDP-RD. Will fix.

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c:
```diff
@@ -62,14 +64,13 @@ _nib_onl_entry_t *_nib_onl_alloc(const ipv6_addr_t *addr, unsigned iface)
 
-    assert(addr != NULL);
     DEBUG("nib: Allocating on-link node entry (addr = %s, iface = %u)\n",
           ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
```
this line will crash if `addr == NULL`

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c:
```diff
@@ -341,6 +343,96 @@ _nib_dr_entry_t *_nib_drl_get_dr(void)
 }
 
+_nib_offl_entry_t *_nib_offl_alloc(const ipv6_addr_t *next_hop, unsigned iface,
+                                   const ipv6_addr_t *pfx, unsigned pfx_len)
```
indent

miri64 replied:
Fix your browser font :P

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c:
```diff
@@ -341,6 +343,96 @@ _nib_dr_entry_t *_nib_drl_get_dr(void)
+    _nib_offl_entry_t *dst = NULL;
+
+    assert((pfx != NULL) && (!ipv6_addr_is_unspecified(pfx)) &&
+           (pfx_len != 0) && (pfx_len <= 128));
```
just taste, but I like `pfx_len > 0` more, because it's in harmony with `px_len <= 128`

miri64 replied:
Dito, don't know what came over me.

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h:
```diff
@@ -427,15 +430,15 @@ _nib_dr_entry_t *_nib_drl_get_dr(void);
-_nib_offl_entry_t *_nib_dst_alloc(const ipv6_addr_t *next_hop, unsigned iface,
-                                  const ipv6_addr_t *pfx, unsigned pfx_len);
+_nib_offl_entry_t *_nib_offl_alloc(const ipv6_addr_t *next_hop, unsigned iface,
+                                   const ipv6_addr_t *pfx, unsigned pfx_len);
```
indent

miri64 replied:
Is fixed here, I know :P

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h:
```diff
@@ -506,16 +504,16 @@ static inline void _nib_dst_remove(_nib_offl_entry_t *nib_offl, uint8_t mode)
                                              const ipv6_addr_t *dst)
 {
-    return _nib_dst_add(next_hop, iface, dst, IPV6_ADDR_BIT_LEN, _DC);
+    assert(next_hop != NULL);
```
please also add `assert(dst != NULL);`

cgundogan replied:
That may be true for now, but what if this code changes with time and possibly by another developer?. IMO these `asserts` also have a documentational character.

cgundogan commented on tests/unittests/tests-gnrc_ipv6_nib/tests-gnrc_ipv6_nib-internal.c:
```diff
@@ -964,6 +965,735 @@ static void test_nib_drl_get_dr__success4(void)
+ * interfaces and then tries to add another.
+ * Expected result: should return NULL
+ */
+static void test_nib_offl_alloc__no_space_left_diff_next_hop_iface(void)
```
I dont' think you need the above two test cases when you introduce this test case that combines both

miri64 replied:
I disagree, checking for behavior different addresses, different interfaces and different addresses *and* different interfaces are disjunct test-cases. One simple mix-up of `&&` and `||` in the implementation would let fail at least one of them.

cgundogan commented on tests/unittests/tests-gnrc_ipv6_nib/tests-gnrc_ipv6_nib-internal.c:
```diff
@@ -964,6 +965,735 @@ static void test_nib_drl_get_dr__success4(void)
+ * tries to add another.
+ * Expected result: should return NULL
+ */
+static void test_nib_offl_alloc__no_space_left_diff_next_hop_iface_pfx(void)
```
here the same, the above two test cases are not really necessary, are they?

miri64 replied:
See my comment above.

cgundogan commented on tests/unittests/tests-gnrc_ipv6_nib/tests-gnrc_ipv6_nib-internal.c:
```diff
@@ -964,6 +965,735 @@ static void test_nib_drl_get_dr__success4(void)
+ * equal to the last.
+ * Expected result: should return not NULL (the last)
+ */
+static void test_nib_offl_alloc__success_duplicate(void)
```
So many test cases that are very similar .. can you collapse them? It is hard to maintain so many test cases in case of changes.

miri64 replied:
There are subtle differences that are not easily parameterized out. Ideally, this API shouldn't change much.

miri64 commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h:
```diff
@@ -506,16 +504,16 @@ static inline void _nib_dst_remove(_nib_offl_entry_t *nib_offl, uint8_t mode)
                                              const ipv6_addr_t *dst)
 {
-    return _nib_dst_add(next_hop, iface, dst, IPV6_ADDR_BIT_LEN, _DC);
+    assert(next_hop != NULL);
```
But it's checked in `_nib_offl_alloc()` (which is called by `_nib_offl_add()`)...

miri64 commented:
Comments addressed.

miri64 commented:
> Found an error here, while debugging NDP-RD. Will fix.

Done: the problem was, if a Prefix List entry was allocated first and a neighbor cache entry to the same interface later, the allocation would create a new entry. The last two commits add a fix and a corresponding test for this.

miri64 commented:
Oops, now the address isn't set ^^ will fix

miri64 commented:
(fixed and added corresponding test)

miri64 commented:
Ping @cgundogan? You promised to review last week :(

cgundogan commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.h:
```diff
@@ -232,8 +237,6 @@ static inline void _nib_onl_set_if(_nib_onl_entry_t *node, unsigned iface)
  *
- * @pre     `(addr != NULL)`.
- *
  * @param[in] addr  An IPv6 address. May not be NULL.
```
May ~~not~~ be NULL.

miri64 commented:
Addressed comments

cgundogan commented:
@miri64 great, thanks for your patience! Please squash the commits and then let's murdock have some fun!

miri64 commented:
Squashed

cgundogan commented:
Murdock spits out some errors here.

miri64 commented:
> Murdock spits out some errors here.

Problem was the amended inline function requested in https://github.com/RIOT-OS/RIOT/pull/7212#discussion_r143117187. Fixed and squashed immediately.

cgundogan commented:
Now murdock is happy (: GO
